### PR TITLE
Implement auto-suspend function for Kobo, which does not have a power management service in kernel

### PR DESCRIPTION
**Pull request by Hzj-jie**
Discussed in #1356 #1565 & #1814.
@Frenzie, let's discuss in this PR.

### CODE REVIEW

**Comment by Hzj-jie** [comment]
@Frenzie regarding your comments,
1. But it would be suspended during suspend. ;-)
I cannot quite get the point, AFAIK, UIManager is still working in suspend mode, so I do not think we should stop auto-suspend procedure to save battery.
2. The user facing setting should be in minutes. And then preferably you could still use e.g. 10.5 minutes instead of 630 seconds, so you could set it to less than a minute while testing.
I did not see a similar configuration has been exported to end users in Kindle, DuoKan or Nickel. So do we really need to make this feature user-facing? My initial plan is to let it configurable, but only via a configuration file change, no user interface.

**Comment by KenMaltby** [comment]
Such a function without User configuration would be very annoying. In fact one of the configurations made available needs to be one that disables the function.

**Comment by Hzj-jie** [comment]
@KenMaltby, to be very clear, this function has a configuration to control the timeout. But as far as I can tell, most of users, include me, do not really care about the timeout value. If it's long enough to avoid to be automatically suspended too often. So I chose 15 min, which seems a common value across different e-reader systems, as the default value.

File: frontend/ui/uimanager.lua [hunk 1]
```diff
@@ -352,6 +360,8 @@ function UIManager:_checkTasks()
     local now_us = now[1] * MILLION + now[2]
     local wait_until = nil
 
+    -- task.action may schedule other events
+    self._task_queue_dirty = false
```
Comment (houqp) [inline frontend/ui/uimanager.lua]: Can you write a unit test for this change?

File: frontend/ui/uimanager.lua [hunk 2]
```diff
@@ -617,6 +627,33 @@ function UIManager:runForever()
...
+            self:scheduleIn(
+                self.last_action_sec + self.auto_suspend_sec - now,
+                self.auto_suspend_action)
+        end
+    end
+    self:_startAutoSuspend()
+end
+
+function UIManager:_startAutoSuspend()
+    self.last_action_sec = util.getsec()
```
Comment (houqp) [inline frontend/ui/uimanager.lua]: To get the second, you can just use `last_action_sec = util.gettime()` or return `util.getrtime()` directly in `util.getsec`. That way, we can avoid creating a new table every time `getsec` is called.

**Comment by houqp** [comment]
> The user facing setting should be in minutes. And then preferably you could still use e.g. 10.5 minutes instead of 630 seconds, so you could set it to less than a minute while testing.

I 100% agree we should use minute for user facing stuff. For this one, if we are going to make it configurable for the majority of the users, we should just make menu item for it in the UI, which uses minute as unit. But I hope we can hide this from the majority of the users, see below. For power users, I think it's fine to leave it as second. I do see how using second as a unit in the reader config can make the handling logic in the code a little bit cleaner.

> Such a function without User configuration would be very annoying. In fact one of the configurations made available needs to be one that disables the function.

This was my first thought ;)

The second thing came into my mind was a UX design philosophy I read awhile ago. I can't recall the exact wording, but it goes something like below:

> Adding yet another configurable option is often an escape. You either don't know, or don't want to think and decide for your users, so you simply add an option. It's way more harder to find out your users needs and add a feature which will "simply work", without any additional customization.

I hope the quote is pretty self-explanatory ;) So let's see if we can come up with a sane default setting that works for the majority of our users. @KenMaltby , what's your concern on not being able to disable this feature? Do you have any use case in mind that needs the reader to be always on?

**Comment by Frenzie** [comment]
@Hzj-jie
> I cannot quite get the point, AFAIK, UIManager is still working in suspend mode, so I do not think we should stop auto-suspend procedure to save battery.

You could be running mprime and it still wouldn't affect battery life while suspended. Otherwise what would be the point of suspend?

> I did not see a similar configuration has been exported to end users in Kindle, DuoKan or Nickel. So do we really need to make this feature user-facing? My initial plan is to let it configurable, but only via a configuration file change, no user interface.

At the very least, you need to be able to set it to `0` (i.e., off). Also I think it should be at least 30 minutes, especially if it's not easily user-configurable. The point is to keep the device from being active overnight (i.e. at least 8 hours, quite likely many more), not to annoy the user by being suspended after a brief lack of touching the device.

@houqp 
> what's your concern on not being able to disable this feature? Do you have any use case in mind that needs the reader to be always on?

It's not about always on, but about control. Turning on and off my monitor manually gave me much better power efficiency without any annoyance. I despise broken-by-design turn-off-is-unplug DisplayPort. I did have automatic monitor stand-by enabled before switching to UHD, but to something like 50 minutes so it wouldn't stay on for hours in the extremely unlikely case I forgot. Over the years I've found 50 minutes to be the minimum in not annoying me by constantly turning off against my will.

> The second thing came into my mind was a UX design philosophy I read awhile ago. I can't recall the exact wording, but it goes something like below:

It's a well-known paradigm, often used to defend a lack of settings. It's only true if you're just sticking in a default without any thought as to whether it's a sane default. For example, I once read the creator of a code editor defending the lack of a font-size setting by saying that you really shouldn't be using small fonts anyway. But setting aside the question as to whether or not he/she was right about small fonts (e.g., on my 1024x600 netbook every pixel counts and a "small font" wouldn't look especially small anyway), the obvious question immediately presents itself: what if you want *larger* fonts?

> @KenMaltby , what's your concern on not being able to disable this feature? Do you have any use case in mind that needs the reader to be always on?

If I put a book on my table to occasionally reference something while I'm working, it doesn't automatically shut itself. (Okay, depending on the binding, whether I'm inside or outside etc. it may do that, but you get the point.) My e-reader is an inferior simulacrum of paper, and most deviations from paper (such as auto-suspend) are for the worse.

One random example is a [game map](http://metroid.retropixel.net/games/metroid3/metroid3_map.gif). Another is a glossary or the notes to something else. Or maybe you're working on a crossword.

Practically speaking, an autosuspend that activates only after hours of inactivity would in almost all imaginable cases never annoy the user. But in the interest of battery life, something in the 30 to 60 minute range should do the trick.

PS What about an option to suspend like in an EPUB without a cover, except without a big message covering up all the text in the center? More like a little Zzz... mark in the corner or something. In fact that way you could be far more aggressive about suspend, activating it immediately after loading the Metroid map.

**Comment by KenMaltby** [comment]
Frenzie hit on several. I also believe that "it should just work" is a worthwhile design goal, but there is also a need for "I should be able to have it work as I want it to".  The two are not mutually exclusive. 
In the early days of the suspend function, my favorite "screensaver" was just a frozen screen, with no message box.  It was easy enough to comment out the "show" command.  To be honest, it probably won't effect me personally, as I can just disable the code myself.  As I mentioned it would be an annoyance.

File: frontend/ui/uimanager.lua (hunk 2 above)
Comment (Hzj-jie) [inline frontend/ui/uimanager.lua]: Sorry, I am not an expert of lua :) I will change koreader-base. But I did not see a util.getrtime function.

**Comment by Hzj-jie** [comment]
+1, totally agree.
@KenMaltby Indeed this function is able to disable, if you set the timeout to a very large value (60*60*24*10 = 864000 or upper, I do believe battery of Kobo is pretty long lasting, but 10 days should be enough to drain out.)

**Comment by KenMaltby** [comment]
I assume that the counting is reset when the device is powered off, restarted, at least.  It would make sense that the counting reset when the auto-suspend takes place.  The disabling would be to not have my use of the device interrupted by an unwanted function. It is not that my use would require it not suspend for any defined amount of time, just that it remain not suspended unless I do the suspending.  That you may want otherwise is fine, but it should be optional for those who do not want the device suspending on its own.  As I said, I personally, can delete or disable any scripting.

File: frontend/ui/uimanager.lua (hunk 2 above)
Comment (houqp) [inline frontend/ui/uimanager.lua]: Sorry, I made a typo, i meant return `util.gettime()` directly in `util.getsec`.  I would suggest we just use `util.gettime()` directly.

**Comment by houqp** [comment]
> > I cannot quite get the point, AFAIK, UIManager is still working in suspend mode, so I do not think we should stop auto-suspend procedure to save battery.
>
> You could be running mprime and it still wouldn't affect battery life while suspended. Otherwise what would be the point of suspend?

In suspend mode, UIManager should be blocked on the select system call and not doing anything until a resume event got generated. Looks like @Hzj-jie already added the code to unschedule auto suspend task :)

Based on the feedback we have so far, I suggest we do the following:

* Set the default timeout to 1h or longer. Eink device should consume very little battery on a static screen anyway, so there is no need to set a small timeout like we do with LCD devices. THB, even my laptop has a 45 minutes auto dim ;)
* For the setting parsing, we should disable auto suspend if `auto_suspend_timeout_seconds` is not set. I can help write a reader setting patch module after this one to push out default on new config during version update. Setting it to a very long time feels like a hack to me and it's better to do it clean from the beginning.

**Comment by Frenzie** [comment]
I feel like not set is default (e.g. 1 hour) and 0 disabled is how I'm used
to config working. Unless default is disabled.

**Comment by Hzj-jie** [comment]
Emm, a conflict, if auto_suspend_timeout_seconds is not set, should we use 1 hour or disable?
My preference is,
if auto_suspend_timeout_seconds <= 0, disable this feature.
if auto_suspend_timeout_seconds is not set, use default value as 1 hour.

As we have discussed before, typical users do not expect the reader will be always on -- also I still believe 15 min is a reasonable idle timeout, as it follows settings of other eink-reader software :) --. So using 1 hour auto-suspend timeout as a default behavior (not set) looks like a good choice.

**Comment by houqp** [comment]
> if auto_suspend_timeout_seconds <= 0, disable this feature.
> if auto_suspend_timeout_seconds is not set, use default value as 1 hour.

Let's do it this way :+1: 

I wonder why the other systems set it to 15 minutes as default. Maybe they just throw a number out there, or maybe they have their own concern. But I just can't think a good reason to do it on eink devices ;) That said, feel free to raise your concern about the 1 hour default timeout here if you have any. We can discuss further.

**Comment by KenMaltby** [comment]
The one hour should cover the "fell asleep while reading" scenario, the most likely one.

**Comment by Hzj-jie** [comment]
As far as I can tell,
1. It's not so common for a regular reader to be stuck at one page for over 15 minutes. You always need to turn page.
2. Not all the e-readers are as battery economic as Kobo. I have two Kindles, 4 and voyage. 4 is fairly good, since it has no front light, I can continually read for some 40 hours when it was brand new. Voyage is bad, I believe it cannot last for 20 hours with front light off, and even shorter with front light on. So 1 hour idle timeout is not so efficient for these devices.

**Comment by Frenzie** [comment]
I'm on board.

**Comment by houqp** [comment]
> 1. It's not so common for a regular reader to be stuck at one page for over 15 minutes. You always need to turn page.

There are times when you would want to have the reader displaying a reference content for a long period of time. For example, having the reader displaying a diagram or paper on the side for reference while implementing an algorithm.

> 2. Not all the e-readers are as battery economic as Kobo. I have two Kindles, 4 and voyage. 4 is fairly good, since it has no front light, I can continually read for some 40 hours when it was brand new. Voyage is bad, I believe it cannot last for 20 hours with front light off, and even shorter with front light on. So 1 hour idle timeout is not so efficient for these devices.

IIRC, @chrox mentioned kindle and pocketbook already have auto suspend "feature" from the native system? Since this is only a KOBO feature for now, I would prefer we don't sacrifice the user experience on KOBO because amazon did not do a good job on controlling power consumption on their devices ;) A static eink screen should be consuming very little power really. If that's not the case, it's likely a bug in the software given they are using similar product from the same manufacturer.

If we decided to roll this out to all platforms later, we can always tune the default setting down then.

**Comment by Frenzie** [comment]
As a regular reader, I just leave my book while going to the
bathroom/answering the door/whatever and expect it to be exactly as I left
it when I come back. If it auto-suspended in the meantime that's not a
disaster, but it'd still be annoying. In principle auto-suspend is still
annoying after doing the same thing overnight, but with such timespans it
becomes a necessary evil.

**Comment by Hzj-jie** [comment]
Sorry, I was busy these days, I will try to polish the change according to our discussion above ASAP.
Besides this change, should we add a 'power users' section in wiki, to explain some hidden configurations, book marks, etc?

**Comment by houqp** [comment]
> Sorry, I was busy these days, I will try to polish the change according to our discussion above ASAP.
> Besides this change, should we add a 'power users' section in wiki, to explain some hidden  configurations, book marks, etc?

This would nice. I am also thinking of bundling the content of wiki into user documentation in KOReader release in the future.

**Comment by houqp** [comment]
Sorry, I missed your new commit because github did not send me any update :( Looks good to me, could you rebase against the latest master?

**Comment by Hzj-jie** [comment]
Definitely.

**Comment by houqp** [comment]
Thanks, can you also write a simple unit test for the change in `_checkTasks()`? Just to make sure we don't have regression in the future in case of refactoring.

**Comment by Hzj-jie** [comment]
Ah, yes, I should do it.

File: frontend/ui/uimanager.lua (hunk 1 above)
Comment (Hzj-jie) [inline frontend/ui/uimanager.lua]: Missed your comment.

**Comment by houqp** [comment]
:+1:

**Pull request merged**